crates/node-optimism/src/evm/execute.rs

```diff
@@ -0,0 +1,743 @@
+    }
+
+    /// Configures an optional inspector stack for debugging.
+    pub fn with_inspector(mut self, inspector: Option<InspectorStack>) -> Self {
+        self.inspector = inspector;
+        self
+    }
+
+    /// Configures the prune modes for the executor.
+    pub fn with_prune_modes(mut self, prune_modes: PruneModes) -> Self {
+        self.prune_modes = prune_modes;
+        self
+    }
+}
+
+impl<EvmConfig> OpExecutorProvider<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = Bytes>,
+{
+    fn op_executor<DB>(&self, db: DB) -> OpBlockExecutor<EvmConfig, DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        OpBlockExecutor::new(
+            self.chain_spec.clone(),
+            self.evm_config.clone(),
+            State::builder().with_database(db).with_bundle_update().without_state_clear().build(),
+        )
+        .with_inspector(self.inspector.clone())
+    }
+}
+
+impl<EvmConfig> ExecutorProvider for OpExecutorProvider<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = Bytes>,
+{
+    type Executor<DB: Database<Error = ProviderError>> = OpBlockExecutor<EvmConfig, DB>;
+
+    type BatchExecutor<DB: Database<Error = ProviderError>> = OpBatchExecutor<EvmConfig, DB>;
+    fn executor<DB>(&self, db: DB) -> Self::Executor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        self.op_executor(db)
+    }
+
+    fn batch_executor<DB>(&self, db: DB) -> Self::BatchExecutor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        let executor = self.op_executor(db);
+        OpBatchExecutor {
+            executor,
+            batch_record: BlockBatchRecord::new(self.prune_modes.clone()),
+            stats: BlockExecutorStats::default(),
+        }
+    }
+}
+
+/// Helper container type for EVM with chain spec.
+#[derive(Debug, Clone)]
+struct OpEvmExecutor<EvmConfig> {
+    /// The chainspec
+    chain_spec: Arc<ChainSpec>,
+    /// How to create an EVM.
+    evm_config: EvmConfig,
+}
+
+impl<EvmConfig> OpEvmExecutor<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = Bytes>,
+{
+    /// Executes the transactions in the block and returns the receipts.
+    ///
+    /// This applies the pre-execution changes, and executes the transactions.
+    ///
+    /// # Note
+    ///
+    /// It does __not__ apply post-execution changes.
+    fn execute_pre_and_transactions<Ext, DB>(
+        &mut self,
+        block: &BlockWithSenders,
+        mut evm: Evm<'_, Ext, &mut State<DB>>,
+    ) -> Result<(Vec<Receipt>, u64), BlockExecutionError>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        // apply pre execution changes
+        apply_beacon_root_contract_call(
+            &self.chain_spec,
+            block.timestamp,
+            block.number,
+            block.parent_beacon_block_root,
+            &mut evm,
+        )?;
+
```

Comment:
maybe some `is_bedrock` check to exit with error is helpful

Reply:
pre bedrock execution is not possible, so this is out of scope here